Replace Unicorn with Puma

skyschub opened the pull request:
This PR primarily replaces Unicorn with Puma. Unicorn has been unmaintained for quite some time, and on the other hand, Puma has been in production on at least Geraspora and Nerdpol for quite some time. 

For podmins that still run the default setup, this will probably result in a significant reduction in memory usage, as Puma is a lot more efficient. (It also doesn't need to be killed regularly...) Unfortunately, this change means some breaking config changes, but oh well. I have included a Changelog entry, so you don't need to come up with something during the merge.

This PR also contains one commit that removes traces of a 12+ years old attempt at creating some federation integration testing, and it updates all binstubs.

skyschub committed:
Remove artifacts from the old Federation integration testing attempt

It's… safe to assume that nobody is using this, as this has been broken
for quite some time.

skyschub committed:
Update binstubs

Newly generated binstubs will check for the string
`This file was generated by Bundler` inside `bin/bundle`, so we'd have
to update that anyway.

Also, there is a non-zero chance the updated `bundle` binstub resolves
some of the setup-specific issues we've seen.

SuperTux88 reviewed (changes requested):
I found two little things.

Also, as podmins shouldn't change `web_workers` anymore, I was also wondering if we should add a config option to set the maximum number of threads, or if the default (5) is fine for all pod sizes? Or for really big pods would that still be the point where they would increase the workers if they need more than 5 threads?

denschub commented on config/eye.rb:
```diff
@@ -14,39 +14,35 @@
 
     if rails_env == "production"
-      start_command "#{unicorn_command} -D"
+      start_command "#{web_command} -D"
```
Right, I missed that. You're right, there is no need to even have a difference here anymore, so I removed the entire env-gate, used the old production setup, but set `daemonize true`.

denschub commented:
Review feedback addressed in the force push.

> I was also wondering if we should add a config option to set the maximum number of threads, or if the default (5) is fine for all pod sizes? Or for really big pods would that still be the point where they would increase the workers if they need more than 5 threads?

So... I don't *think so*. Puma will auto-scale its thread pool with a default range of `0:5`. Pods with spikes in traffic will see an increase in memory usage, that might result in funky square-wave memory curves, like that one:

<img width="637" alt="Screenshot 2022-09-09 at 04 13 24" src="https://user-images.githubusercontent.com/344777/189257769-b86dafc1-a3dc-4706-ad8a-c9d052ae93f4.png">

There, I had a bit of a (self-induced) load spike, and Puma kept the extra threads around for more than a day, just in case, before eventually releasing it again.

I don't think that the maximum of 5 is a problem. Puma will not scale up if the memory pressure is high, so there should not be ever a case where we're risking OOM. Large pods with *lots* of requests will see higher memory usage than smaller pods, but that's expected. Likewise, I don't think podmins will ever need *more* than 5 threads. I've never seen Geraspora exceed three threads, and I'm fairly sure that a pod would be completely on fire because of database latency before Puma scaling becomes a problem. 

So, I don't think we need to make threads configurable. And, writing that, I realize I might just have made a case for removing the workers pref altogether - I don't think there is ever a case where a podmin would need more than one worker. And if a podmin does, then they're able to change the Puma config directly.

SuperTux88 commented:
> And, writing that, I realize I might just have made a case for removing the workers pref altogether - I don't think there is ever a case where a podmin would need more than one worker. And if a podmin does, then they're able to change the Puma config directly.

Yes, I think we should remove the `web_workers` setting then. And podmins can also set the max threads by setting a `MAX_THREADS` env var without editing the config, and if that then really isn't enough anymore, then yes, I think they can change the puma config themselves.

skyschub committed:
Replace Unicorn with Puma

… and drop the single_process_mode. See the included Changelog entry
for full details on what this change means.

denschub commented:
Setting removed - and Changelog changed accordingly - in [another force-push](https://github.com/diaspora/diaspora/compare/68e407a1f8da527abc95d03943994b41034beccc..97cfc80a1fe6b712d15c13081cc938d3650fbdb1).

denschub commented:
(Note to bystanders: rspec tests are unhappy in this PR. But that's not my fault. It's September 9th, which means the profile that's used to verify that the Birthday Notification System isn't spamming everyone, ... [has its birthday today](https://github.com/diaspora/diaspora/issues/8394)...)

SuperTux88 reviewed (approved):
Looks good now :cookie: and I'm just ignoring the broken tests now ;)

SuperTux88 commented:
Thanks for the finding with the binstubs, who knew that they needed to be re-generated, this isn't documented in the upgrade guide. But as this fixes the bundler issue a few podmins faced (verified locally), I backported it as 9075dfa470d473e345758b14f3704ede15c451de and released it as [0.7.18.1](https://github.com/diaspora/diaspora/releases/tag/v0.7.18.1). :cookie:
